lucene/core/src/java/org/apache/lucene/search/TopDocs.java

```diff
@@ -260,11 +291,12 @@ private static TopDocs mergeAux(Sort sort, int start, int size, TopDocs[] shardH
       }
       if (shard.scoreDocs != null && shard.scoreDocs.length > 0) {
         availHitCount += shard.scoreDocs.length;
-        queue.add(new ShardRef(shardIDX, setShardIndex == false));
+        queue.add(new ShardRef(shardIDX));
       }
     }
 
     final ScoreDoc[] hits;
+    final boolean unsetShardIndex = shardHits[0].scoreDocs[0].shardIndex == -1;
```

Comment:
this is prone to out-of-bounds exceptions, you should find the first TopDocs instance that has a non-empty ScoreDoc[] if any